docs/contrib/keycloak.md

```diff
@@ -0,0 +1,148 @@
+# Keycloak Identity Provider for REMS
+
+Using docker, docker-compose, and Keycloak, you can roll your own local IdP for use with REMS.
+
+## Quick-ish Start
+
+1. **Setting up the Docker files**
+    1. Create a `Dockerfile` for your keycloak. To start, it can be as simple as the following:
+    ```
+    FROM quay.io/keycloak/keycloak:11.0.2
+    ```
+    2. Create a `docker-compose.yaml` file for REMS and your Keycloak IdP. Mount your custom REMS config to `/rems/config/config.edn`. Example:
+    ```
+    version: "3.7"
+    services:
+    rems:
+        mem_reservation: 200m
+        mem_limit: 500m
+        ports:
+        - "127.0.0.1:3001:3000"
+        volumes:
+        - ./services/rems/simple-config.edn:/rems/config/config.edn
+        depends_on:
+        - db
+        - keycloak
+    db:
+        image: postgres:9.6
+        environment:
+        POSTGRES_USER: rems
+        POSTGRES_PASSWORD: remspassword
+        mem_reservation: 30m
+        mem_limit: 150m
+        ports:
+        - "127.0.0.1:5432:5432"
+    keycloak:
+        environment:
+        KEYCLOAK_USER: admin
+        KEYCLOAK_PASSWORD: admin
+        ports:
+        - "3002:8080"
+    ```
+
+2. **Preparing a Keycloak Realm**
+    1. Build and run keycloak with `docker-compose up keycloak`.
+    2. Login to keycloak at `http://keycloak:8080/auth/admin` with the admin username and password, set to `admin` and `admin` in this (very insecure) example. These can be set as environment variables, as demonstrated in the `docker-compose.yaml` example above.
+    3. Add an identity provision `realm` to Keycloak by mousing over `Master` near the top-left of the page, and clicking `Add realm`. A realm can be thought of as a single vault containing a set of users, credentials, groups, clients, security settings, etc. When authenticating with a realm, the client is only able to authenticate within that realm. Realms can be partially exported and imported by clicking `Import`/`Export` near the bottom-left of the page, although exporting users is more complicated. For this example, the realm will be named `rems-idp`.
+    4. Add a client for REMS, by navigating to `Clients` on the left side of the page and clicking `Create`. Set the `Client Protocol` to `openid-connect`. Clients are allowed to authenticate users within the Keycloak realm that they are bound to. For this example, the Client ID will be `rems-client`.
```

Comment:
Could we perhaps modify these instructions to show the correct value as not the last word but maybe in the very first sentence? Scanning through the instructions would be a bit faster I think.

Reply:
There's no correct or incorrect value here, you can set the client name to anything. I'm just stating what I set it to in this example so that it's easier to see where certain strings come from in the config files for REMS, docker-compose, etc.

Reply:
You are right. What I meant is that perhaps there could be a "table" of settings to use, even if they are just example values. And they would then match the different steps of the instructions, and overall you get a working setup without much thinking. Then you can tune them later to what you wish. And even so I would put the example values as first just to be clear as what the target is. Maybe phrasing like this:

Next we will set up Keycloak with the following settings

| Keycloak | Setting | Value |
|-----------|----------|---------|
| Realm | Name | rems-idp |
| Client  | Client Protocol | openid-connect |
| ... | ... | ...|

1. ...
2. ...
3. **Add an identity provision `realm` to Keycloak named `rems-idp`.**
You can do that by mousing over `Master` near the top-left of the page, and clicking `Add realm`. 
A realm can be thought of as a single vault containing a set of users, credentials, groups, clients, security settings, etc. When authenticating with a realm, the client is only able to authenticate within that realm. Realms can be partially exported and imported by clicking `Import`/`Export` near the bottom-left of the page, although exporting users is more complicated.